From bf1b138fd3e041c0ce144693161f8d403febf5c6 Mon Sep 17 00:00:00 2001
From: Emilian Peev <epeev@google.com>
Date: Mon, 19 Sep 2022 15:55:52 -0700
Subject: [PATCH 2/4] camera2 vndk: Improve error handling

Check for binder status when removing the camera
service listener.
Additionally check for any invalid extended availability
callbacks.

Bug: 247615866
Test: adb shell
/data/nativetest64/vendor/ACameraNdkVendorTest/ACameraNdkVendorTest

Change-Id: Ia043934288d8c5e5584034a08de24c998468039d
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 camera/ndk/ndk_vendor/impl/ACameraManager.cpp | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/camera/ndk/ndk_vendor/impl/ACameraManager.cpp b/camera/ndk/ndk_vendor/impl/ACameraManager.cpp
index 77c934abbe..bb4ef5634a 100644
--- a/camera/ndk/ndk_vendor/impl/ACameraManager.cpp
+++ b/camera/ndk/ndk_vendor/impl/ACameraManager.cpp
@@ -165,7 +165,10 @@ CameraManagerGlobal::~CameraManagerGlobal() {
     Mutex::Autolock _l(mLock);
     if (mCameraService != nullptr) {
         mCameraService->unlinkToDeath(mDeathNotifier);
-        mCameraService->removeListener(mCameraServiceListener);
+        auto stat = mCameraService->removeListener(mCameraServiceListener);
+        if (!stat.isOk()) {
+            ALOGE("Failed to remove listener to camera service %s", stat.description().c_str());
+        }
     }
     mDeathNotifier.clear();
     if (mCbLooper != nullptr) {
@@ -475,6 +478,10 @@ void CameraManagerGlobal::CallbackHandler::onMessageReceivedInternal(
                 ALOGE("%s: Cannot find camera callback fp!", __FUNCTION__);
                 return;
             }
+            if (cb == nullptr) {
+                // Physical camera callback is null
+                return;
+            }
             found = msg->findPointer(kContextKey, &context);
             if (!found) {
                 ALOGE("%s: Cannot find callback context!", __FUNCTION__);
-- 
2.34.1

